domain-models-runtime/src/main/java/org/folio/rest/persist/PgUtil.java

```diff
@@ -439,4 +498,137 @@ public static void deleteById(String table, String id,
+    }
+    List<ModifierSet> sortIndexes = cqlSortNode.getSortIndexes();
+    if (sortIndexes.size() != 1) {
+      throw new SQLOptimizationException("sort index is missing");
+    }
+    ModifierSet modifierSet = sortIndexes.get(0);
+    if (! modifierSet.getBase().equals(column)) {
+      throw new SQLOptimizationException("column name does not match");
+    }
+    String ascDesc = getAscDesc(modifierSet);
+    cql = cqlSortNode.getSubtree().toCQL();
+    String lessGreater = "";
+    if(ascDesc.equals("DESC") )
+    {
+      lessGreater = ">" ;
+    } else {
+      lessGreater = "<";
+    }
+    preparedCql.getCqlWrapper().setQuery(cql);
+    String tableName = PostgresClient.convertToPsqlStandard(tenantId)
+        + "." + preparedCql.getTableName();
+    String where = preparedCql.getCqlWrapper().getField().toSql(cql).getWhere();
+    // If there are many matches use a full table scan in title sort order
+    // using the title index, but stop this scan after OPTIMIZED_SQL_SIZE index entries.
+    // Otherwise use full text matching because there are only a few matches.
+    //
+    // "headrecords" are the matching records found within the first OPTIMIZED_SQL_SIZE records
+    // by stopping at the title from "OFFSET OPTIMIZED_SQL_SIZE LIMIT 1".
+    // If "headrecords" are enough to return the requested "LIMIT" number of records we are done.
+    // Otherwise use the full text index to create "allrecords" with all matching
+    // records and do sorting and LIMIT afterwards.
+    String sql =
+        " WITH "
+      + " headrecords AS ("
+      + "   SELECT jsonb, lower(f_unaccent(jsonb->>'" + column + "')) AS title FROM " + tableName
+      + "   WHERE (" + where + ")"
+      + "     AND lower(f_unaccent(jsonb->>'" + column + "'))" + lessGreater
+      + "             ( SELECT lower(f_unaccent(jsonb->>'" + column + "'))"
+      + "               FROM " + tableName
+      + "               ORDER BY lower(f_unaccent(jsonb->>'" + column + "')) " + ascDesc
+      + "               OFFSET " + size + " LIMIT 1"
+      + "             )"
+      + "   ORDER BY lower(f_unaccent(jsonb->>'" + column + "')) " + ascDesc
+      + "   LIMIT " + limit + " OFFSET " + offset
+      + " ), "
+      + " allrecords AS ("
+      + "   SELECT jsonb, lower(f_unaccent(jsonb->>'" + column + "')) AS title FROM " + tableName
+      + "   WHERE (" + where + ")"
+      + "     AND (SELECT COUNT(*) FROM headrecords) < " + limit
+      + " )"
+      + " SELECT jsonb, title,  0                                 AS count"
+      + "   FROM headrecords"
+      + "   WHERE (SELECT COUNT(*) FROM headrecords) >= " + limit
+      + " UNION"
+      + " (SELECT jsonb, title, (SELECT COUNT(*) FROM allrecords) AS count"
+      + "   FROM allrecords"
+      + "   ORDER BY title " + ascDesc
+      + "   LIMIT " + limit + " OFFSET " + offset
+      + " )"
+      + " ORDER BY title " + ascDesc;
+
+      logger.info("optimized SQL generated from CQL: " + sql);
+
+    return sql;
```

Comment:
The method is not complete because it does not execute the sql.
See the code from mod-inventory-storage: https://github.com/folio-org/mod-inventory-storage/blob/v15.1.0/src/main/java/org/folio/rest/impl/InstanceStorageAPI.java#L265

Reply:
I added the execution in a later commit